linera-views/src/batch.rs

```diff
@@ -373,63 +373,60 @@ impl DeletePrefixExpander for MemoryContext<()> {
+    type Iter: BatchValueWriter<Self>;
 
-    /// Returns an owning iterator on the simplified batch.
+    /// Returns an owning iterator over the values in the batch.
     fn into_iter(self) -> Self::Iter;
 
-    /// Returns the total number of entries of the simplified batch
+    /// Returns the total number of entries in the batch.
     fn len(&self) -> usize;
 
-    /// Returns the total number of bytes of the simplified batch
-    fn bytes(&self) -> usize;
+    /// Returns the total number of bytes in the batch.
+    fn num_bytes(&self) -> usize;
 
-    /// Returns the overhead size of the simplified batch
+    /// Returns the overhead size of the batch.
     fn overhead_size(&self) -> usize;
 
-    /// Appends the iterator to the simplified batch
-    fn try_append(
-        &mut self,
-        iter: &mut Self::Iter,
-        value_size: &mut usize,
-    ) -> Result<bool, bcs::Error>;
-
-    /// Returns the serialization and clears the simplified batch
-    fn to_bytes(&mut self) -> Result<Vec<u8>, bcs::Error>;
-
-    /// Adds an individual delete operation
+    /// Adds an individual delete operation.
     fn add_delete(&mut self, key: Vec<u8>);
 
-    /// Adds an individual insert of (key,value)
+    /// Adds an individual insert of (key,value).
     fn add_insert(&mut self, key: Vec<u8>, value: Vec<u8>);
 
-    /// Gets the next size of iterator
-    fn next_value_size(
-        &self,
-        value_size: usize,
-        iter: &mut Self::Iter,
-    ) -> Result<usize, bcs::Error>;
-
-    /// Creates a simplified batch from an existing batch
+    /// Creates a simplified batch from an existing batch.
     async fn from_batch<S: DeletePrefixExpander + Send + Sync>(
         store: S,
         batch: Batch,
     ) -> Result<Self, S::Error>;
 
-    /// Returns true if the Simplified batch is empty
+    /// Returns true if the batch is empty.
     fn is_empty(&self) -> bool {
         self.len() == 0
     }
 }
 
-/// The iterator over the simplified batch
-pub trait SimplifiedBatchIter: Send + Sync {
-    /// Returns the number of remaining entries in the iterator.
-    fn remaining_len(&self) -> usize;
+/// An iterator-like object that can write values one by one to a batch while updating the
+/// total size of the batch.
+pub trait BatchValueWriter<Batch>: Send + Sync {
+    /// Returns true if there is no more values to write.
+    fn is_empty(&self) -> bool;
+
+    /// Writes the next value (if any) to the batch and updates the batch size accordingly.
+    ///
+    /// Returns false if there was no value to write.
+    fn write_next_value(
+        &mut self,
+        batch: &mut Batch,
+        batch_size: &mut usize,
+    ) -> Result<bool, bcs::Error>;
+
+    /// Computes the batch size that we would obtain if we wrote the next value (if any)
+    /// without consuming the value.
+    fn next_batch_size(&mut self, batch: &Batch, batch_size: usize) -> Result<usize, bcs::Error>;
```

Comment:
If I understand correctly, `batch_size` never includes the bytes needed to encode the list lengths, but the returned value does? We could add that to the doc string here.

Reply:
Yes good point.

Reply:
Changed to `next_value_size` (only returns the size of the next value)

Reply:
Making a separate PR (#1502)